Userland/Libraries/LibJS/Console.cpp

```diff
@@ -511,24 +512,128 @@ ThrowCompletionOr<Value> ConsoleClient::logger(Console::LogLevel log_level, Mark
-    } else {
-        // 6. Otherwise, perform Printer(logLevel, Formatter(args)).
+    // 5. Otherwise, perform Printer(logLevel, Formatter(args)).
+    else {
         auto formatted = TRY(formatter(args));
         TRY(printer(log_level, formatted));
     }
 
-    // 7. Return undefined.
+    // 6. Return undefined.
     return js_undefined();
 }
 
 // 2.2. Formatter(args), https://console.spec.whatwg.org/#formatter
 ThrowCompletionOr<MarkedVector<Value>> ConsoleClient::formatter(MarkedVector<Value> const& args)
 {
-    // TODO: Actually implement formatting
-    return args;
+    auto& vm = m_console.vm();
+    auto& realm = *vm.current_realm();
+    MarkedVector<Value> result { vm.heap() };
+
+    // 1. If args’s size is 1, return args.
+    if (args.size() == 1)
+        return args;
+
+    // 2. Let target be the first element of args.
+    auto target = (!args.is_empty()) ? TRY(args.first().to_string(vm)) : "";
+
+    // 3. Let current be the second element of args.
+    auto current = (args.size() > 1) ? args[1] : js_undefined();
+
+    // 4. Find the first possible format specifier specifier, from the left to the right in target.
+    auto find_specifier = [](StringView target) -> Optional<StringView> {
+        size_t start_index = 0;
+        while (start_index < target.length()) {
+            auto maybe_index = target.find('%');
+            if (!maybe_index.has_value())
+                return {};
+
+            auto index = maybe_index.value();
+            if (index + 1 >= target.length())
+                return {};
+
+            switch (target[index + 1]) {
+            case 'c':
+            case 'd':
+            case 'f':
+            case 'i':
+            case 'o':
+            case 'O':
+            case 's':
+                return target.substring_view(index, 2);
+            }
+
+            start_index = index + 1;
+        }
+        return {};
+    };
+    auto maybe_specifier = find_specifier(target);
+
+    // 5. If no format specifier was found, return args.
+    if (!maybe_specifier.has_value())
+        return args;
+
+    // 6. Otherwise:
+    {
+        auto specifier = maybe_specifier.release_value();
+        Optional<Value> converted;
+
+        // 1. If specifier is %s, let converted be the result of Call(%String%, undefined, « current »).
+        if (specifier == "%s"sv) {
+            converted = TRY(call(vm, realm.intrinsics().string_constructor(), js_undefined(), current));
+        }
+        // 2. If specifier is %d or %i:
+        else if (specifier.is_one_of("%d"sv, "%i"sv)) {
+            // 1. If Type(current) is Symbol, let converted be NaN
+            if (current.is_symbol()) {
+                converted = js_nan();
+            }
+            // 2. Otherwise, let converted be the result of Call(%parseInt%, undefined, « current, 10 »).
+            else {
+                converted = TRY(call(vm, realm.intrinsics().parse_int_function(), js_undefined(), current, Value { 10 }));
```

Comment:
I am SO happy with now nicely this whole cleanup turned out :^)

Reply:
Yup, you LibJS folks have done a great job at making the API nicer recently!